### ISSUE TITLE
Submission pruning, field columns deleted?

### ISSUE THREAD

**Issue by nickdunn** [verbb/formie#589]
**Description**
This may be a red herring, but I'm chasing down all avenues right now.

We have a site that has been running well for months. Seemingly randomly overnight Craft started erroring all on requests. The logs showed that a number of columns had been dropped from the `content` table which was causing entry queries to fail. The field references continued to exist elsewhere in the database, so it seemed just these columns dropped.

The fields in question are from a normal section modelling People (first name, last name, email, job title etc.) which coincidentally map to some Formie form fields we have too.

The first activity in the Craft logs I can see yielding the error was a `Submissions->pruneDataRetentionSubmissions` event which we run on an hourly cron via the CLI. 

```
Next yii\db\Exception: SQLSTATE[42S22]:   Column not found: 1054 Unknown column 'content.field_email' in 'field   list'
```

<img width="829" alt="Screenshot 2021-09-29 at 17 24 34" src="https://user-images.githubusercontent.com/39560/135309787-6894a03c-06b2-4bd9-a22d-a378380e299d.png">

Is there any chance at all Formie could have a hand in deleting columns from the `content` table?

**Additional info**
- Plugin version: latest
- Craft version: latest

**Comment by engram-design** [comment]
So Formie shouldn't touch your `content` table, instead all submissions use their own `fmc_*` tables. It should certainly not be dropping columns.

We really need to figure out the ID of the submission element throwing this error. It's possible this submission belongs to a form that no longer exists, which might cause this error. Otherwise, I'm not sure why its resolving it's content table to `content` instead of `fmc_*`.

Would there be any chance of a database dump? Feel free to get in touch via `web at verbb.io`.

**Comment by nickdunn** [comment]
Thanks Josh. 

Further down the logs I see:

>  #20 /var/www/craft/vendor/craftcms/cms/src/base/ApplicationTrait.php(1477): craft\services\Gc->run()

Could this be triggering `deleteOrphanedFields` which erroneously dropped columns? As you say, unsure why it would do this to the main `content` table and not a Formie table though.

Database dumps of before and after sent by email.

**Comment by engram-design** [comment]
I did think that's the only real possibility, but the content table is supposed to be set before deleting the field, so it should only delete the column from the `fmc_` table.

Thanks for the database, I'll dive into that to see if I can reproduce.

Just to confirm what commands you're running on an ongoing basis via Cron? Just the `./craft formie/gc/prune-data-retention-submissions` command?

**Comment by nickdunn** [comment]
```
craft queue/run
craft blitz/cache/refresh-expired
craft seomatic/sitemap/generate
craft formie/gc/prune-data-retention-submissions
```

**Comment by engram-design** [comment]
Looks like your suspicions were right, this is a Formie issue and a failing of how we implemented `deleteOrphanedFields()`.

When there are two fields with the exact same handle in Formie and the global scope and the Formie field has been marked as orphaned (which itself shouldn’t happen typically, but it can) this error occurs. So there’s a very specific set of circumstances that’ve lead to this, which is probably why it hasn’t cropped up yet. Despite having the correct field to delete, Craft’s `deleteField()` has handling in it that removes the content column from the content table. We can change this to the Formie content table, however this isn’t possible because we can't determine which one it is.

Being an orphaned field means the form it was owned by has gone. The form is what stores the content table fields should use, and with that gone, there’s no way for our `deleteOrphanedFields()` function to know where the fields should be removed from. As such, we can't tell Craft which Formie content table to focus on when deleting the field, so it uses the default one (`content`). We still want to delete the field, just not have it cleanup the content table.

As I mentioned, this only crops up if you have two fields with the same handle. My apologies for the issue.

Should be fixed in [1.4.19](https://github.com/verbb/formie/releases/tag/1.4.19).

**Comment by nickdunn** [comment]
Oof, nasty bug. Thanks for debugging this so quickly :-)

Thankfully Azure's MySQL backups are ace and I was able to restore to within a few minutes of the issue occurring so we didn't lose anything overall.
